### PR TITLE
Valid actions ctypes

### PR DESCRIPTION
**Pull request by mhauskn**
Creates a ctypes alternative for the valid action search.

### CODE REVIEW

File: jericho/jericho.py [hunk 1]
```diff
@@ -522,8 +527,29 @@ def get_valid_actions(self, use_object_tree=True):
         interactive_objs  = self._identify_interactive_objects(use_object_tree=use_object_tree)
         best_obj_names    = self._score_object_names(interactive_objs)
         candidate_actions = self.act_gen.generate_actions(best_obj_names)
-        diff2acts         = self._filter_candidate_actions(candidate_actions)
-        valid_actions     = [max(v, key=utl.verb_usage_count) for v in diff2acts.values()]
+        if use_ctypes:
+            state = self.get_state()
+            candidate_str = (";".join(candidate_actions)).encode('utf-8')
+            valid_str = (' '*len(candidate_str)).encode('utf-8')
```
Comment (MarcCote) [inline jericho/jericho.py:533]: In the C function `self.frotz_lib.filter_candidate_actions` a `; ` is appended after each valid action, i.e.
> valid_actions[v_idx++] = ';';

This means `valid_str` length is short by 1 character (in the case where all candidate actions are valid). Here's a quick fix:
> valid_str = (' '*(len(candidate_str) + 1)).encode('utf-8')

File: jericho/jericho.py [hunk 2]
```diff
@@ -522,8 +527,29 @@ def get_valid_actions(self, use_object_tree=True):
         interactive_objs  = self._identify_interactive_objects(use_object_tree=use_object_tree)
         best_obj_names    = self._score_object_names(interactive_objs)
         candidate_actions = self.act_gen.generate_actions(best_obj_names)
-        diff2acts         = self._filter_candidate_actions(candidate_actions)
-        valid_actions     = [max(v, key=utl.verb_usage_count) for v in diff2acts.values()]
+        if use_ctypes:
```
Comment (MarcCote) [inline jericho/jericho.py:530]: I think this if statement would be better off within `_filter_candidate_actions`.

**Review by MarcCote — CHANGES_REQUESTED** [review]
@mhauskn I thought I'd try to parallelize that computation. Here's what I got
```
Rom roms/zork1.z5 Python 10.9 Ctypes 4.9 Speedup 123.1%
Rom roms/zork1.z5 Python 10.9 Para 4.0 Speedup 173.9%
Rom roms/zork1.z5 Python 10.9 ParaC 2.4 Speedup 358.7%

Rom roms/enchanter.z3 Python 60.9 Ctypes 16.2 Speedup 276.5%
Rom roms/enchanter.z3 Python 60.9 Para 18.1 Speedup 236.4%
Rom roms/enchanter.z3 Python 60.9 ParaC 7.7 Speedup 692.6%
```
where `Para` is a parallel version without using your C function, and ParaC is a hybrid version that split the action candidates into chunks and tests them with your C function.

Do you want me to make a PR to your branch (which will update this PR automatically)?

**Comment by mhauskn** [comment]
Thank you Marc for the corrections - I've implemented both in the latest push! The speedups from your parallelization look quite nice - can you make a PR to include them?

A benchmark which would be valuable with the parallelization is the multi-environment use case. E.g. using a vectorized environment setup like DRRN as well as Redis DB-world-state-hashing. In this case, I suspect the speedups from parallelization will be less than the single-environment case, but hopefully better than no parallelization. Let me know what you think.

**Comment by MarcCote** [comment]
Done. See my PR to your branch.

Regarding "vectorized environment setup like DRRN", that's totally an issue since VecEnv uses daemon processes and those processes can't have children, i.e. can't use `mp.Pool` to compute the valid actions.

I found a workaround using [concurrent.futures.ProcessPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ProcessPoolExecutor) but that was way slower (probably some issue with the Python GIL or something). :(

**Comment by MarcCote** [comment]
Did some additional benchmarking.
Using
`concurrent.futures.ProcessPoolExecutor` for the outer parallelization loop (i.e., VecEnv) but using the `multiprocessing.Pool` for Jericho works great.

**Comment by MarcCote** [comment]
I've updated PR #51.

File: frotz/src/interface/frotz_interface.c [hunk 3]
```diff
@@ -1699,3 +1700,127 @@ void test() {
...
+
+  act = strtok(candidate_actions, ";");
+  while (act != NULL)
+  {
+    // Add a newline and termination to act
+    act_newline = malloc(strlen(act) + 2);
+    strcpy(act_newline, act);
+    strcat(act_newline, "\n");
+
+    // Step: Code is copied due to inexplicable segfault when calling step() directly; Ugh!
```
Comment (MarcCote) [inline frotz/src/interface/frotz_interface.c:1759]: Do you think it was the same issue as with the `opcode` (not sure why it would be the case, though)?

Comment (mhauskn) [reply frotz/src/interface/frotz_interface.c:1750]: I don't think so - it was a segfault and happens on games like Zork which never had the illegal opcode issue.

**Comment by mhauskn** [comment]
This PR looks ready to go from my end. One question is whether we would want to enabled the ctypes + parallelization of valid actions by default? I suspect that it should help in most cases - what do you think @MarcCote?

**Comment by MarcCote** [comment]
I had a similar thought. I would say yes, make the parallel (which uses ctypes) version the default. If we see that it breaks at some point, we can revise this decision.

File: frotz/src/interface/frotz_interface.c (hunk 3 above)
Comment (MarcCote) [reply frotz/src/interface/frotz_interface.c:1760]: ok

**Review by MarcCote — APPROVED** [review]
You can go ahead, change the default for get_valid_actions to `use_ctypes=True` and `use_parallel=True`. Then merge this PR.

**Pull request merged**